File: backend/internal/domain/abuse.go
package domain

import (
	"context"
	"encoding/json"
	"time"
)

const (
	ReportTargetUser    = "user"
	ReportTargetListing = "listing"
	ReportTargetMessage = "message"
	ReportTargetReview  = "review"

	ReportReasonSpam                 = "spam"
	ReportReasonFraud                = "fraud"
	ReportReasonHarassment           = "harassment"
	ReportReasonInappropriateContent = "inappropriate_content"
	ReportReasonPersonalData         = "personal_data"
	ReportReasonOther                = "other"

	ReportStatusNew = "new"
)

// AbuseReport is an immutable user complaint plus its moderation state. The
// evidence is produced by the server from the reported object, never accepted
// from the reporting client.
type AbuseReport struct {
	ID             int64
	ReporterUserID int32
	ReportedUserID int32
	TargetType     string
	TargetID       int64
	Reason         string
	Details        string
	Status         string
	Evidence       json.RawMessage
	Source         string
	AppVersion     string
	IPAddress      string
	UserAgent      string
	CreatedAt      time.Time
}

// ReportTarget is the repository-verified owner and snapshot of a reported
// entity. Message targets are additionally authorized against conversation
// membership before this value is returned.
type ReportTarget struct {
	UserID   int32
	Evidence json.RawMessage
}

type CreateAbuseReport struct {
	ReporterUserID int32
	TargetType     string
	TargetID       int64
	Reason         string
	Details        string
	Source         string
	AppVersion     string
	IPAddress      string
	UserAgent      string
}

type BlockedUser struct {
	UserID    int32
	Name      string
	AvatarURL string
	BlockedAt time.Time
}

type BlockedUsersPage struct {
	Items  []BlockedUser
	Total  int64
	Limit  int32
	Offset int32
}

// UserBlockState describes the pair-level restriction from the viewer's
// perspective. BlockedByMe is intentionally directional: clients may offer an
// unblock action only to the user who created the block, while the other side
// receives neutral copy without learning who initiated it.
type UserBlockState struct {
	Blocked     bool `json:"blocked"`
	BlockedByMe bool `json:"blocked_by_me"`
}

// UserBlockChecker is the narrow read contract consumed by chat and booking.
// Keeping it separate from AbuseRepository prevents those use cases from
// depending on complaint creation or block-list mutations.
type UserBlockChecker interface {
	IsBlockedBetween(ctx context.Context, firstUserID, secondUserID int32) (bool, error)
	BlockState(ctx context.Context, viewerUserID, otherUserID int32) (UserBlockState, error)
}

// AbuseRepository owns both complaint evidence collection and pair-level
// blocks. CreateReport must enforce its daily limit transactionally so it
// remains correct with multiple API replicas.
type AbuseRepository interface {
	CreateReport(ctx context.Context, report CreateAbuseReport, maxPerDay int32) (AbuseReport, error)
	BlockUser(ctx context.Context, blockerUserID, blockedUserID int32) (BlockedUser, error)
	UnblockUser(ctx context.Context, blockerUserID, blockedUserID int32) error
	ListBlockedUsers(ctx context.Context, blockerUserID, limit, offset int32) (BlockedUsersPage, error)
	IsBlockedBetween(ctx context.Context, firstUserID, secondUserID int32) (bool, error)
	BlockState(ctx context.Context, viewerUserID, otherUserID int32) (UserBlockState, error)
}

File: mobile/src/lib/api/abuse.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';

import { api } from '@/lib/api/client';
import { chatKeys } from '@/lib/api/chat';

export type ReportTargetType = 'user' | 'listing' | 'message' | 'review';
export type ReportReason =
  | 'spam'
  | 'fraud'
  | 'harassment'
  | 'inappropriate_content'
  | 'personal_data'
  | 'other';

export interface UserBlockState {
  blocked: boolean;
  blocked_by_me: boolean;
}

export interface BlockedUser {
  user_id: number;
  name: string;
  avatar_url?: string;
  blocked_at: string;
}

export interface BlockedUsersPage {
  items: BlockedUser[];
  total: number;
  limit: number;
  offset: number;
}

export interface CreateReportInput {
  target_type: ReportTargetType;
  target_id: number;
  reason: ReportReason;
  details?: string;
}

export interface ReportReceipt {
  id: number;
  target_type: ReportTargetType;
  target_id: number;
  reason: ReportReason;
  status: 'new' | string;
  created_at: string;
}

export const abuseKeys = {
  all: ['abuse'] as const,
  userBlockState: (userID: number) => [...abuseKeys.all, 'user-block-state', userID] as const,
  blockedUsers: () => [...abuseKeys.all, 'blocked-users'] as const,
};

export function createReport(input: CreateReportInput): Promise<ReportReceipt> {
  return api.post<ReportReceipt>('/api/v1/reports', {
    ...input,
    details: input.details?.trim() ?? '',
  });
}

export function useCreateReport() {
  return useMutation({ mutationFn: createReport });
}

export function fetchUserBlockState(userID: number): Promise<UserBlockState> {
  return api.get<UserBlockState>(`/api/v1/users/${userID}/block-state`);
}

export function useUserBlockState(userID: number | undefined, enabled = true) {
  return useQuery({
    queryKey: abuseKeys.userBlockState(userID ?? 0),
    queryFn: () => fetchUserBlockState(userID as number),
    enabled: enabled && userID != null && Number.isFinite(userID) && userID > 0,
    staleTime: 15_000,
  });
}

export function blockUser(userID: number): Promise<BlockedUser> {
  return api.post<BlockedUser>(`/api/v1/users/${userID}/block`);
}

export function unblockUser(userID: number): Promise<void> {
  return api.delete<void>(`/api/v1/users/${userID}/block`);
}

export function useBlockUser(userID: number) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: () => blockUser(userID),
    onSuccess: () => {
      queryClient.setQueryData<UserBlockState>(abuseKeys.userBlockState(userID), {
        blocked: true,
        blocked_by_me: true,
      });
      queryClient.invalidateQueries({ queryKey: abuseKeys.blockedUsers() });
      queryClient.invalidateQueries({ queryKey: chatKeys.all });
    },
  });
}

export function useUnblockUser(userID: number) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: () => unblockUser(userID),
    onSuccess: () => {
      queryClient.setQueryData<UserBlockState>(abuseKeys.userBlockState(userID), {
        blocked: false,
        blocked_by_me: false,
      });
      queryClient.invalidateQueries({ queryKey: abuseKeys.blockedUsers() });
      queryClient.invalidateQueries({ queryKey: chatKeys.all });
    },
  });
}

export function useUnblockUserMutation() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (userID: number) => unblockUser(userID),
    onSuccess: (_data, userID) => {
      queryClient.setQueryData<UserBlockState>(abuseKeys.userBlockState(userID), {
        blocked: false,
        blocked_by_me: false,
      });
      queryClient.invalidateQueries({ queryKey: abuseKeys.blockedUsers() });
      queryClient.invalidateQueries({ queryKey: chatKeys.all });
    },
  });
}

export function fetchBlockedUsers(limit = 100, offset = 0): Promise<BlockedUsersPage> {
  return api.get<BlockedUsersPage>(`/api/v1/me/blocked-users?limit=${limit}&offset=${offset}`);
}

export function useBlockedUsers(enabled = true) {
  return useQuery({
    queryKey: abuseKeys.blockedUsers(),
    queryFn: () => fetchBlockedUsers(),
    enabled,
    staleTime: 15_000,
  });
}
